version-upgrade-operator/README.md

```diff
@@ -0,0 +1,484 @@
+
+## Rough idea
+
+We create a new operator (tentative name "version-upgrade-operator") and a set of CRDs which will provide and
+act on information about where to discover the latest available version tags, when to update them and on which
+resources.
+
+The new CRs will allow to provide a reusable schedule object, version discovery source and covering many
+target CRs with just 1 instance. The new CRs will be separate from the controller being actually responsible
+for application deployment (like helm-controller or app-operator).
+
+We design the version setting for App CRs and HelmReleases, but the solution is not CR specific. Using the new
+CRDs, an operator will be responsible for setting target CRs' selected property, like the `spec.version` of
+App CR, to an automatically determined value. The value will be discovered in a Helm charts repository and
+applied to a target CR only when time-based restrictions are met.
+
+### Image tags and registries discovery
+
+For the solution to work, we have to discover tags available for specific image (repository) in a registry. We
+figured out that instead of implementing our own logic for discovering tags, we can reuse the existing logic
+present in Flux controllers.
+
+#### Reusing Flux components
+
+Flux has version discovery implemented in a few places:
+
+- `HelmRelease/HelmRepository/HelmChart` level: these are the objects that drive actual Helm Chart deployment.
+  We want to edit them to point to a limited set of possible versions to make them work with the operator
+  we're designing
+- `OCIRepository` which allows to discover and filter tags of artifacts in an OCI repository; it is always
+  downloading the most recent matching artifact
+- `ImageRepository` and `ImagePolicy` are objects normally used by
+  [`ImageAutomationController`](https://fluxcd.io/flux/components/image/), but we can use them for easy
+  discovery of what's available in a registry.
+
+To not interfere with how upgrades might be already configured for existing `HelmReleases`, we will use a
+dedicated `ImageRepository` and `ImagePolicy` objects for discovery process related to automatic upgrades.
+
+##### How this would work?
+
+To configure an automated update for an App CR, we need answers to 3 key questions: 1) what should be
+upgraded; 2) when it should be upgraded; 3) which version should be used. The following steps have to be
+performed by a user and the new operator (this assumes that there's no GitOps interfering, deployment with
+GitOps is described below):
+
+1. The user deploys an application, either by using `App CR` and a related `Catalog` or `HelmRelease` and
+   `HelmRepository`
+   - at this point, Flux or another related deployment operator are already performing Helm Chart deployment
+1. The user has to define an automated version upgrade policy
+   - if it doesn't exist or a user wants to own the object, he creates an `ImageRepository` object, that
```

Comment:
It's a bit confusing to me why it keeps referring to images when the targets are actual Apps or Helm charts.